### PR TITLE
WIP: First steps for enabling the add API in pyWATTS

### LINKED ISSUE #28: Implement new API
For enabling integration of sktime and pyWATTS-pipeline, we need a new API that does not rely on the call dunder. Thus, this issue is about building such an API. It relies in previous discussion in
* #1 
* https://github.com/sktime/sktime/pull/3026#issuecomment-1239648856
* https://github.com/sktime/sktime/issues/2654

As a result of these discussions, we aim to provide an API, which enables to construct a pipeline using `pipeline.add(...)`. To enable such an API beside the current one, the following things need to be implemented:

* Inside of the pipeline, the transformers/modules need to be cloned to be compliant to the sklearn approach
* The steps of the pipeline should be considered as parameters of the pipeline. I.e., it should be set via the set_params method. 
  * To be compliant with sklearn again, we should also provide a possibility to construct the complete pipeline with it's steps using the constructor (set_params and init should have the same parameters [see sklearn](https://scikit-learn.org/stable/developers/develop.html) 
* To enable an easy maintenance of the three APIs, they should use the same logic under the hood. My current proposal is a control flow similar to the following for the three APIs:
  * Constructor Based API: 
    * Pipeline.\_\_init__
    * Pipeline._add() for each step to be added
  * Pipeline.add API (Imperative API): 
    * Pipeline.add(): It calls set_params with all parameters of the previous Pipeline.add() calls
    * Pipeline.set_params: It calls Pipeline._add() for each step to be added (analog to the constructor)
  * Functional API: It transforms the parameters in a way that it can use Pipeline.add() internally.

### PR DESCRIPTION
**Pull request by benHeid**
## Description
This issue introduces two new APIs for sktime. These are the imperative API and the API based on the constructor. I try to reduce the maintenance effort by reusing most of the methods for all APIs. In the corresponding jupyter notebook a more detailed description is given how the new APIs work under the hood. 

The change of the API requires further changes as:
* the identification of the steps. We now identify the steps by its name instead of an extra id. Thus, in the steps, base, base_summary, etc. several corresponding changes has to be made. 
* the step_information does not contains the step anymore instead it contains the step name
* The load and store functionality has to be revised since the ids etc. are removed and the identification changed.

Furthermore the new API now ensures that
* The modules inside the pipeline are cloned. 
* If a new module is added the pipeline is constructed again. Therfore, the pipeline stores the parameters of all previously add statements.

## Related Issues
closes #28 

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos)
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] Did you update the CHANGELOG

## PR review
Anyone in the community is free to review the PR once the tests have passed.

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones to the PR so it can be classified

### CODE REVIEW

**Comment by benHeid** [comment]
# TODOs:
* [ ] The control flow is not correct
  * The pipeline.add currently does everything, however it should only collect previous calls and pass them to set_params
* [ ] The tests are failing
* [ ] Tests for new APIs are missing

File: pywatts_pipeline/core/pipeline.py [hunk 1]
```diff
@@ -57,16 +60,110 @@ class Pipeline(BaseTransformer):
     :type path: str
     """
 
-    def __init__(self, path: Optional[str] = ".", name="Pipeline"):
+    def __init__(self, path: Optional[str] = ".", name="Pipeline", steps=None):
```
Comment (benHeid) [inline pywatts_pipeline/core/pipeline.py:63]: The constructor API needs to look differently. We need to distinguish between structure and modules here, since sklearn assumes that if parameters are deep cloned stay the same. However, in case if a module is added multiple times to the pipeline, there are two entries in the list steps that should contain the identical module. Howevever because of the clone they are not identical any more

File: pywatts_pipeline/core/pipeline.py [hunk 2]
```diff
@@ -57,16 +60,110 @@ class Pipeline(BaseTransformer):
...
+        ])
+        return self._add(self._pipeline_construction_informations)[-1]
+
+    def _add(self, steps):
+        self.steps = {}
+        self.start_steps = {}
+        from pywatts_pipeline.core.steps.step_factory import StepFactory
+        step_informations = []
+        for transformer, name, input_edges, add_params in steps:
+            if id(transformer) not in self.est_dict:
```
Comment (benHeid) [inline pywatts_pipeline/core/pipeline.py:128]: Do not use id(object) instead use the object directly as key. 

id(object) may cause problems if the object is destroyed. Then the id can be assigned to another object.

Comment (benHeid) [reply pywatts_pipeline/core/pipeline.py:128]: See the following example
```
b = "test"
print(id(b))
del b
c = "test"
print(id(c))
```
